src/tree/splitter.ts

```diff
@@ -0,0 +1,206 @@
+  impurity_left: number = Number.POSITIVE_INFINITY
+  impurity_right: number = Number.POSITIVE_INFINITY
+  left_value: int[] = []
+  right_value: int[] = []
+  found_split = false
+}
+
+export class Splitter {
+  kMinSplitDiff_: number
+  feature_data_: number[][]
+  label_data_: int[]
+  criterion_: Criterion
+  start_: int
+  end_: int
+  min_samples_leaf_: int
+  max_features_: int
+  feature_order_: int[]
+  shuffle_features_: boolean
+
+  sample_map_: SampleData[]
+  n_samples_total_: int
+  n_features_: int
+  constructor(
+    feature_data: number[][],
+    label_data: int[],
+    min_samples_leaf: int,
+    impurity_measure: ImpurityMeasure,
+    max_features: int,
+    samples_subset: int[] = []
+  ) {
+    this.feature_data_ = feature_data
+    this.label_data_ = label_data
+    this.n_features_ = feature_data[0].length
+    this.min_samples_leaf_ = min_samples_leaf
+    this.max_features_ = max_features
+    this.shuffle_features_ = max_features < this.n_features_
+    this.sample_map_ = []
+    this.start_ = 0
+    this.end_ = 0
+    this.kMinSplitDiff_ = 1e-8
+    if (samples_subset.length === 0) {
+      this.n_samples_total_ = feature_data.length
+      for (let i = 0; i < this.n_samples_total_; i++) {
+        this.sample_map_.push({ current_feature_value: 0, sample_number: i })
+      }
+    } else {
+      this.n_samples_total_ = samples_subset.length
+      for (let i = 0; i < this.n_samples_total_; i++) {
+        this.sample_map_.push({
+          current_feature_value: 0,
+          sample_number: samples_subset[i]
+        })
+      }
+    }
+    this.criterion_ = new Criterion(impurity_measure, label_data)
+    this.feature_order_ = []
+    for (let i = 0; i < this.n_features_; i++) {
+      this.feature_order_.push(i)
+    }
+    this.resetSampleRange(0, this.n_samples_total_)
+  }
+
+  resetSampleRange(start: int, end: int) {
+    this.start_ = start
+    this.end_ = end
+    this.criterion_.init(start, end, this.sample_map_)
+  }
+
+  splitNode() {
+    let current_split = new Split()
+    let best_split = new Split()
+    let current_impurity_improvement = Number.NEGATIVE_INFINITY
+    let best_impurity_improvement = Number.NEGATIVE_INFINITY
+    let current_feature_num = 0
+    let current_feature = 0
+    current_split.found_split = false
+    if (this.shuffle_features_) {
+      this.feature_order_ = shuffle(this.feature_order_)
+    }
+
+    while (current_feature_num < this.max_features_) {
+      current_feature = this.feature_order_[current_feature_num]
+
+      // Copies feature data into sample map
+      for (let i = this.start_; i < this.end_; i++) {
+        this.sample_map_[i].current_feature_value =
+          this.feature_data_[this.sample_map_[i].sample_number][
+            current_feature
+          ]
+      }
+      this.criterion_.reset()
+
+      this.sample_map_ = quickSort(
+        this.sample_map_,
+        this.start_,
+        this.end_ - 1,
+        'current_feature_value'
+      )
```

Comment:
If you turned `sample_map` into an `Int32Array`, You could also use:
```js
sample_map.subarray(start, end).sort(
  (i,j) => feature_data_[i][current_feature]
         - feature_data_[j][current_feature]
)
```

Reply:
Bro this is gold! I'm definitely gonna do this update, and get rid of all the quicksort code.